### ISSUE TITLE
ES.45 for loop

### ISSUE THREAD

**Issue by magisterRab** [isocpp/CppCoreGuidelines#895]
Instead
constexpr int month_count = 12;   // months are numbered 1..12

for (int m = first_month; m <= month_count; ++m)   // better
    cout << month[m] << '\n';

I suggest 
constexpr int month_count = 12;   // months are numbered 1..12

for (int m = first_month; m < first_month + month_count; ++m)   // better
    cout << month[m] << '\n';

**Comment by MikeGitb** [comment]
I completely agree, that this example should be fixed. 

- For one, `first_month` doesn't naturally pair with `month_count` - the limit should be `last_month` - so this code only works, if you know that `first_month` is 1, which  devalues the use of the constant in the first place.

- Second, in 90% of the casees `<= [...]_count/size` is a warning signal that there is either a bug or something fishy is going on. So I usually try to avoid such constructs, even if they are correct, so I don't get distracted.

The second problem I see here however (and which doesn't get fixed by this proposal) is that - unless the type of `month` is a container/range with non-zero based indexing - this:

    for (int m = first_month; m <= month_count; ++m) // better
        cout << month[m] << '\n';

has most likely a different semantic than 

    for (auto m : month)
        cout << m << '\n';

(and in the case of an `std::(unordered_)map`, the latter version wouldn't even work). 

Not saying that such containers aren't used in real code (In fact, I wrote a `Months` class that pretty much works like that), but I think such non-trivial details should either be avoided or explicitly noted in the examples

**Comment by tlanc007** [comment]
I agree this example is unfortunate.  The use of months in the example isn’t showing a very good way of dealing with months.  For example a better way would be to have months as an enumeration and the size could be determined from that.

Would something like the following be any better for a drop-in replacement for this section? 

##### Reason
Unlike variables, an unnamed constant doesn’t describe the role it plays in the expression.  Unnamed constants embedded in expressions are often hard to understand as one has to interpret its meaning.  Additionally within the same expression or a set of expressions the same literal value may not even mean the same thing.

##### Example
```
// Don't do this
    auto my_val {5 * 5 * 3};
    std::string my_str {"ab"};
    my_str += std::string (5 - my_str.length (), '*');
    std::array <int, 5> my_array {};
```

```
// more descriptive 
    constexpr auto width {5ul};
    constexpr auto length {5ul};
    constexpr auto height {3ul};
    auto volume {width * length * height};
    
    std::string my_str {"ab"};
    constexpr auto pad_length {5ul};
    constexpr auto pad_token {'*'};
    my_str += std::string (pad_length - my_str.length () , pad_token);
    
    constexpr auto my_array_size {5ul};
    std::array <int, my_array_size> my_array {};
```

##### Enforcement

Flag literals in code. Give a pass to `0`, `1`, `nullptr`, `\n`, `""`, and others on a positive list.

**Comment by AndrewPardoe** [comment]
Jonathan, would you like to take a stab at improving the existing example?

**Comment by jwakely** [comment]
Thanks for the suggestion. I've fixed this via a smaller change which I think resolves the OP's problem with the loop bounds. The point of the example is to improve on the bad example by using named constants, which I think it does satisfactorily.

I agree this rule could be fleshed out in more detail, possibly by adding @tlanc007's suggestion as a second example (in addition to the existing one). That **Reason** is useful (but I'd also like to see some mention of other reasons to avoid magic numbers, such as not needing to change every occurrence of a magic number if the code is refactored). I think any such improvements can be done via a new issue and/or pull request though.
